posix.mak, win32.mak: Ensure `clean` target removes all generated files and directories

WebDrake opened the pull request:
These patches correct some oversights in the `clean` targets declared by the makefiles.  These should ensure that all generated files are removed, not just the application binaries.

Note that this would automatically be dealt with if `clean` just removed the `$(ROOT_OF_THEM_ALL)` directory.  This option has been avoided so as to avoid changing the existing `clean` behaviour, since presumably there is some perceived risk associated with removing directories recursively.  However, it may be worth returning to this decision in future as (if it could be done safely) it would greatly simplify declaration and maintenance of the `clean` target.

**Warning:** as a non-Windows dev I have not tested the patch to `win32.mak`, so someone may want to try this out in anger before accepting the PR ;-)

CyberShadow commented on posix.mak:
```diff
@@ -59,7 +59,8 @@ DOC_TOOLS = \
 TEST_TOOLS = \
-    $(ROOT)/rdmd_test
+    $(ROOT)/rdmd_test \
+    $(ROOT)/tests_extractor
```
`tests_extractor` cannot be built with just DMD, so this isn't right.

WebDrake replied:
Can you explain more what you see as the problem here?  From my side the point is simply that `clean` should remove all generated files (and this ensures that).

CyberShadow replied:
See the target using `TEST_TOOLS` below:

```
 $(TOOLS) $(DOC_TOOLS) $(CURL_TOOLS) $(TEST_TOOLS): $(ROOT)/%: %.d
$(DMD) $(DFLAGS) -of$(@) $(<)
```

WebDrake replied:
Ouch!  Must say that it's weird then that the CI doesn't catch this.  Is the explicit `$(ROOT)/tests_extractor` target being chosen in preference to the wildcard-y one?

CyberShadow replied:
No idea, actually.

WebDrake replied:
I realized I missed a warning output by `make` about the shadowing of the target definition.  Fixing.

WebDrake commented:
Removed the first patch and reworked the changes to `posix.mak`.  The `TEST_TOOLS` definition is now unchanged.

BTW can I confirm: _is_ there an objection to `clean` just removing the `generated` dir and all its contents?  It would arguably be simpler to maintain.  The question is whether it's considered too risky to do.

CyberShadow commented:
> BTW can I confirm: is there an objection to clean just removing the generated dir and all its contents?

No, IIRC that's what the DMD makefile does.

WebDrake commented:
OK, then I will rewrite accordingly.  It's much simpler that way.

WebDrake commented:
Updated to use `rm -rf` on the directory in which all generated files are placed.

WebDrake commented on posix.mak:
```diff
@@ -94,7 +94,7 @@ install: $(TOOLS) $(CURL_TOOLS) $(ROOT)/dustmite
 
 clean:
-	rm -f $(ROOT)/dustmite $(TOOLS) $(CURL_TOOLS) $(DOC_TOOLS) $(TAGS) *.o $(ROOT)/*.o
+	rm -rf $(ROOT_OF_THEM_ALL)
```
Minor note here: I wasn't sure what the `$(TAGS)` stuff was, since it doesn't appear to have any definition within the makefile itself.  Should that be re-added?

Same question applies to the `win32.mak` changes.

CyberShadow replied:
BTW, `-f` usually isn't needed with `-r`, probably better to just use `rm -R` without `-f`, like the DMD makefile.

CyberShadow commented on posix.mak:
```diff
@@ -94,7 +94,7 @@ install: $(TOOLS) $(CURL_TOOLS) $(ROOT)/dustmite
 
 clean:
-	rm -f $(ROOT)/dustmite $(TOOLS) $(CURL_TOOLS) $(DOC_TOOLS) $(TAGS) *.o $(ROOT)/*.o
+	rm -rf $(ROOT_OF_THEM_ALL)
```
Could you please also rename the `ROOT_OF_THEM_ALL` target to `GENERATED` and `ROOT` to `G`, for consistency with the DMD makefile, and so that we intentionally break any possible scripts that were overriding these variables to `.` or some other directory that should not be deleted?

WebDrake replied:
The DMD makefile is actually the exception here, both druntime and phobos makefiles use `ROOT_OF_THEM_ALL`.

CyberShadow commented on win32.mak:
```diff
@@ -72,7 +72,7 @@ $(ROOT)\changed.exe : changed.d
 
 clean :
-	del $(TARGETS) $(TAGS)
+	del $(ROOT_OF_THEM_ALL)
```
`del` doesn't delete directories, instead use `rmdir /s /q`.

WebDrake commented:
Updated in line with feedback:

```diff
diff --git a/posix.mak b/posix.mak
index cd6bef0..05ddc85 100644
--- a/posix.mak
+++ b/posix.mak
@@ -15,8 +15,8 @@ $(shell [ ! -d $(DMD_DIR) ] && git clone --depth=1 https://github.com/dlang/dmd
 include $(DMD_DIR)/src/osmodel.mak
 
 # Build folder for all binaries
-ROOT_OF_THEM_ALL = generated
-ROOT = $(ROOT_OF_THEM_ALL)/$(OS)/$(MODEL)
+GENERATED = generated
+ROOT = $(GENERATED)/$(OS)/$(MODEL)
 
 # Set DRUNTIME name and full path
 ifeq (,$(findstring win,$(OS)))
@@ -94,7 +94,7 @@ install: $(TOOLS) $(CURL_TOOLS) $(ROOT)/dustmite
        cp $^ $(INSTALL_DIR)/bin
 
 clean:
-       rm -rf $(ROOT_OF_THEM_ALL)
+       rm -r $(GENERATED)
 
 $(ROOT)/tests_extractor: tests_extractor.d
        mkdir -p $(ROOT)
diff --git a/win32.mak b/win32.mak
index 4a2f606..4d4b829 100644
--- a/win32.mak
+++ b/win32.mak
@@ -30,8 +30,8 @@ SCP=$(CP)
 
 DFLAGS=-O -release
 
-ROOT_OF_THEM_ALL = generated
-ROOT = $(ROOT_OF_THEM_ALL)\windows\32
+GENERATED = generated
+ROOT = $(GENERATED)\windows\32
 
 TARGETS=       $(ROOT)\dman.exe \
        $(ROOT)\rdmd.exe \
@@ -72,7 +72,7 @@ $(ROOT)\changed.exe : changed.d
        $(DMD) $(DFLAGS) -of$@ changed.d
 
 clean :
-       del $(ROOT_OF_THEM_ALL)
+       rmdir /s /q $(GENERATED)
 
 detab:
        $(DETAB) $(SRCS)
```

wilzbach commented:
Regarding the dmd Makefile - it's the more "modern" one though it's really sad that Andrei is blocking a unification of the Makefiles
https://github.com/dlang/dmd/pull/7476

WebDrake commented:
Well, what convinced me was @CyberShadow's excellent point about the risk to downstream users' build scripts.  I do not want to be the person responsible for other people losing their hard disk (although hopefully the sort of overrides described would only be done on temporary CI instances...).

Good luck with trying to clean up the build scripts!  Although given the complexities involved, it may ultimately be better that each individual repo is master of its own makefile design ...
